Subject: [PATCH] Restart persistent process if start timeout.

Sometimes bad applications may occupied all binder threads
of system server awhile, then persistent process may be
killed due to background ANR, and meet process start timeout
because no avaiable binder thread to attach.

After binder threads are free, system becomes normal but
the persistent process does not restart, and even it started
by someone use its component, it will have no persistent flag.

diff --git a/services/core/java/com/android/server/am/ActivityManagerService.java b/services/core/java/com/android/server/am/ActivityManagerService.java
index 58c2ba4..0e677d8 100644
--- a/services/core/java/com/android/server/am/ActivityManagerService.java
+++ b/services/core/java/com/android/server/am/ActivityManagerService.java
@@ -6989,6 +6989,9 @@
                 Slog.w(TAG, "Unattached app died before broadcast acknowledged, skipping");
                 skipPendingBroadcastLocked(pid);
             }
+            if (app.persistent && !app.isolated) {
+                addAppLocked(app.info, null, false, null /* ABI override */);
+            }
         } else {
             Slog.w(TAG, "Spurious process start timeout - pid not known for " + app);
         }
